Signalling duplicates in store_key

tobirolinew opened the pull request:
Thanks for this utility, it has been very helpful in my projects.

This pull request adds a flag to subroutine `store_key` to handle the case when the key is already in the hashmap.
If a duplicate is found, the routine now returns `-2`.
This makes it easier to check for and handle duplicates directly.

The patch has been tested and runs correctly.

Would you be able to add this improvement to the project?

Thanks for considering this potential update.

tobirolinew committed:
Update ffhash_inc.f90

It would be useful to have a flag in subroutine "store_key" for the case when the key is already present in the hashmap. I've been using "-2" in the modified version of your code.

jannisteunissen commented:
Thanks for providing this pull request! I have already merged it, and have just made a few more changes: the code now detects negative values (not just -1) as being an error, and I improved the formatting of your commit a bit.

jannisteunissen commented:
@tobirolinew After thinking a bit more about this, I guess it depends a bit on the application whether storing something under the same key twice should give an error. For example, in Python it is fine to redefine something in a dictionary. What would you think about setting a parameter in the hash datastructure that controls whether an error is returned or whether the existing value is overwritten?

tobirolinew commented:
@jannisteunissen Thanks for implementing these changes. 
It is indeed good if a duplicate can be interpreted in two different ways by the hash structure. 
Perhaps it would be useful to have two different procedures for this purpose: 
(a) `store_key()` could be used to add the key even it is redundant, while
(b) `insert_if_absent()` [or a function with similar name] could be an alternative inserter 
     for the non-redundant addition.
Checking an extra condition, like the value of a global variable, could be somewhat slower,
but maybe not so drastic, because you are already checking two logical expressions there:
```
          if (.not. bucket_deleted(h, i) .and. &
               keys_equal(h%keys(i), key)) then
               i = -2
               return
          end if
```
Anyway, I'd let you decide which option you prefer, and I'll adapt to it.

Another thing you might consider. When I compile my code with debug flags, I got the following remark:

```
src/vendor/ffhash/ffhash_inc.f90(367): remark #7712: This variable has not been used.   [H]
  subroutine reset(h)
```

Is that variable truly unused? As far as I remember correctly, I tried to eliminate it, but something went wrong with it.

tobirolinew commented:
I'd have another minor idea which you might consider at some point (absolutely not urgent). 
It could be useful to allow the module name to be set via an `FFH_MOD_NAME` macro. 
That way the `module … end module` wrapper could live directly inside `ffhash_inc.f90`.
For example, instead

`module m_hash_a 
implicit none 
#define FFH_KEY_TYPE integer 
#define FFH_VAL_TYPE character(len=15) 
#define FFH_VAL_IS_STRING 
#include "ffhash_inc.f90" 
end module m_hash_a`

we could write

```
**#define FFH_MOD_NAME m_hash_a** 
#define FFH_KEY_TYPE integer 
#define FFH_VAL_TYPE character(len=15) 
#define FFH_VAL_IS_STRING 
#include "ffhash_inc.f90" 
```

jannisteunissen commented:
I have just pushed a couple commits:

- f9d3b8ea50f396741d0ec1b8510d9c2d5735d7e7
- a4e58af411f615b55576d0999b352772a7497902

On my system, there is no noticeable performance of the new parameter `existing_key_is_error` in the hash table.

The old reset was using the `intent(out)` attribute to completely reset a hash table to its initial state. I have now changed that so that the parameters (load factor and `existing_key_is_error`) are preserved. I think the warning should also disappear now.

tobirolinew commented:
Thank you, @jannisteunissen, for adding this option — it is very helpful, 
especially when using derived types as keys, where avoiding extra copies 
helps prevent memory leaks.

To simplify the process of defining a hashmap for derived types, I've created 
a small typekey helper (see the `ffhash_typekey_helper.f90` and `test.f90` files 
in [helper.zip](https://github.com/user-attachments/files/21886616/helper.zip))
It can use any array- or scalar-like variables as key values (even if they belong to 
subtypes, like in the test case).
The helper’s hash function relies on your `MurmurHash3_x86_32` 
subroutine, so I moved the last `#endif` of `ffhash_inc.f90` in front of
`pure integer(int32) function rotl32(x, r)` [see my related comment 
in `ffhash_inc.f90`].

If you find the `ffhash_typekey_helper.f90` macro useful, feel free 
to include it in your library (with full right to modify).

From my perspective, it would be useful if `ffhash_typekey_helper.f90` 
supported 64-bit hash values.
In my application, I currently use the following simple function:

```
      !*****************************************************************
         pure integer(int64) function hash_function(key)
      !*****************************************************************
         implicit none
         type(FFH_CUSTOM_TYPE), intent(in) :: key
         integer :: i
         integer(int64) :: seed,c,elem
         c=int(z'9e3779b97f4a7c15',8)
         seed=0_8
         do i=1,size(key%FFH_KEY_FIELD)
            elem=int(key%FFH_KEY_FIELD(i),8)
            seed=ieor(seed,elem+c+ishft(seed,6)+ishft(seed,-2))
         end do
         hash_function=seed
         end
```
Do you think this could also be achieved using MurmurHash3?

jannisteunissen commented:
Thanks for your feedback! I had a quick look but do not immediately understand why the flattened arrays are necessary. And the point is thus that you want to hash only part of an object? (The original code simply computes a hash of the whole derived type)

I don't have much time to work on this now, but 64 bit hash values could indeed be nice. I did port a 128 bit MurmurHash3 function at some point (https://github.com/jannisteunissen/murmur3-fortran). But for 64 bit I would have to explore a bit.

tobirolinew commented:
Thanks for your reply.

You are right, the expensive flattening process is not needed. 
I just wanted to use it to treat scalar and array fields together.
Now I added an `FFH_SCALAR_KEY_FIELD` preprocessor variable
to handle scalar fields differently: [helper.zip](https://github.com/user-attachments/files/21919168/helper.zip).
The purpose of this helper would be to simplify the use of `ffhash` 
for derived types which have scalar/array comparables ("key fields").
This "split-module" scheme, where one needs to insert the preprocessing 
step inside the module body, might be a bit tricky to use in some cases. 

I can work on a 64-bit Murmur version and let you know if I managed to 
come up with something reasonable.

jannisteunissen commented:
Do you mean "pointer-like" components in the custom type, such as an allocatable array? I am not sure how that works in the original implementation, I guess the pointer address is hashed.

And I assume you want a 64 bit hash function to 64 bit indexing, which would be necessary for more than a few billion entries?

tobirolinew commented:
I just thought of a value-based hash scheme. 
I use `ffhash` within one of my custom types, where I have an 
integer coordinate vector (`lpo%coors`) guiding the hashing process:

```
!***********************************************************************
      module mod_lpo     ! lpo := lattice point
!***********************************************************************
      use iso_fortran_env
      implicit none
      type :: lpo
         integer, allocatable :: coors(:)
      end type
      !*****************************************************************
      #define FFH_KEY_TYPE type(lpo)
      #define FFH_CUSTOM_KEYS_EQUAL
      #define FFH_CUSTOM_HASH_FUNCTION
      #include "ffhash_inc.f90"
      !*****************************************************************
         pure logical function keys_equal(a,b)
      !*****************************************************************
         type(lpo), intent(in) :: a,b
         integer :: i
         keys_equal=.false.
         do i=1,size(a%coors)
            if(a%coors(i).ne.b%coors(i)) return
         end do
         keys_equal=.true.
         end
      !*****************************************************************
         pure integer*8 function hash_function(key)
      !*****************************************************************
         implicit none
         type(lpo), intent(in) :: key
         integer :: i
         integer*8 :: seed,c,elem
         c=int(z'9e3779b97f4a7c15',8)
         seed=0_8
         do i=1,size(key%coors)
            elem=int(key%coors(i),8)
            seed=ieor(seed,elem+c+ishft(seed,6)+ishft(seed,-2))
         end do
         hash_function=seed
         end
      end

```
This nicely collects the different `coors` vectors via your `ffhash` macro, 
even for billions of lattice points (I often have very large grids, hence my 
need for 64-bit hashers).
The `ffhash_typekey_helper.f90` is just a generalization of the module above, 
for arbitrary scalars/arrays to treat as keys.

tobirolinew commented:
I think I managed to extend `ffhash` to `int64` by using your 128-bit hasher:

```
#ifndef FFH_CUSTOM_HASH_FUNCTION
  pure function hash_function(key) result(hash)
    FFH_KEY_ARG, intent(in)   :: key
    integer(FFH_INT_KIND)     :: hash
    integer(int32), parameter :: seed = 42
#ifdef FFH_KEY_IS_STRING
#ifdef FFH_ENABLE_INT64
    integer(int32)            :: hash128(4)
    integer(int64)            :: h1, h2
    call MurmurHash3_x64_128(key, len_trim(key), seed, hash128)
    h1 = transfer(hash128(1:2), h1)
    h2 = transfer(hash128(3:4), h2)
    hash = ieor(h1, h2)
#else
    call MurmurHash3_x86_32(key, len_trim(key), seed, hash)
#endif
#else
    integer, parameter        :: n_bytes = ceiling(0.125d0*storage_size(key))
    character(len=n_bytes)    :: buf
#ifdef FFH_ENABLE_INT64
    integer(int32)            :: hash128(4)
    integer(int64)            :: h1, h2
    call MurmurHash3_x64_128(transfer(key, buf), n_bytes, seed, hash128)
    h1 = transfer(hash128(1:2), h1)
    h2 = transfer(hash128(3:4), h2)
    hash = ieor(h1, h2)
#else
    call MurmurHash3_x86_32(transfer(key, buf), n_bytes, seed, hash)
#endif
#endif
  end function hash_function
```

See https://github.com/jannisteunissen/ffhash/compare/master...tobirolinew:ffhash:master,
which contains the changes introduced (including four new optional preprocessor variables).
You can find a description in commit `652322f`.
In the other commit (that is, `f5787ea`), I've added two examples to `example_multiple_tables.f90`,
which demostrates the use of the int64 precision.

If you find these changes useful (any of them can be reverted if you disagree with it), 
then I can submit a new pull request.

tobirolinew commented:
I think I managed to extend `ffhash` to `int64` by using your 128-bit hasher:

```
#ifndef FFH_CUSTOM_HASH_FUNCTION
!
!*** BEGIN default hash function
  pure function hash_function(key) result(hash)
    FFH_KEY_ARG, intent(in)   :: key
    integer(FFH_INT_KIND)     :: hash
    integer(int32), parameter :: seed = 42
#ifdef FFH_KEY_IS_STRING
    hash = default_hash_core(key, seed)
#else
    integer, parameter     :: n_bytes = ceiling(0.125d0*storage_size(key))
    character(len=n_bytes) :: buf
    buf = transfer(key, buf)
    hash = default_hash_core(buf, seed)
#endif
  end function hash_function
!*** END default hash function
!
!*** BEGIN helper function (to avoid code duplication)
  pure function default_hash_core(buf, seed) result(hash)
    character(len=*), intent(in) :: buf
    integer(int32),   intent(in) :: seed
    integer(FFH_INT_KIND)        :: hash
#ifdef FFH_ENABLE_INT64
    integer(int32) :: hash128(4)
    integer(int64) :: h1, h2
    call MurmurHash3_x64_128(buf, len(buf), seed, hash128)
    h1 = transfer(hash128(1:2), h1)
    h2 = transfer(hash128(3:4), h2)
    hash = ieor(h1, h2)
#else
    call MurmurHash3_x86_32(buf, len(buf), seed, hash)
#endif
  end function default_hash_core
!*** END helper function
!
#endif
```

[Commit bf26903](https://github.com/jannisteunissen/ffhash/commit/bf269031fa0b12c6f4260f17a61d6584e2a00bb5) contains the main modifications, you can find a small description 
about what has changed.
In [commit 12d8a19](https://github.com/jannisteunissen/ffhash/commit/12d8a1967df6199a155f0a26fb17b6501d47fc80), I've added two examples to `example_multiple_tables.f90`, 
which demostrates the use of the `int64` precision.
The other minor commits are just to add `iso_fortran_env` due to `int32` and `int64`.

If you find these changes useful (any of them can be reverted if you disagree with it), 
then I can submit a new pull request. 
Note that these modifications are backward compatible with your previous examples.

jannisteunissen commented:
That is looking quite good already! I have a few small questions (partially for myself):

- In your code, why did you change to uppercase `FFH_T`?
- I think it is good to have as few preprocessor statements as possible, which is why I did not include a `FFH_EXISTING_KEY_IS_ERROR`.
- The seed for a 64 bit hash function could also be 64 bit I guess

Another point to consider (for myself): According to xxhash (see https://github.com/Cyan4973/xxHash) there are considerably more efficient hash functions than murmur3. However, xxhash itself seemed a bit difficult to port to Fortran. I had a look at a few others, and think I can port spookyhash pretty easily. I will set up a small benchmark to check the performance against murmur3 today.

tobirolinew commented:
Thanks! Let me address these points:

> * In your code, why did you change to uppercase `FFH_T`?

I think it is more convenient to set the `FFH_T` name, instead of aliasing like 
`use m_hash_b, ffhb_t => ffh_t`.

Combining `ffhash_inc.f90` with my helper (`ffhash_typekey_inc.f90`), 
I can use the following settings:

      #define FFH_MOD_NAME mod_hashmap_mytype
      #define FFH_CUSTOM_MODULE mod_mytype
      #define FFH_CUSTOM_TYPE mytype
      #define FFH_KEY_FIELD coors
      #define FFH_KEY_FIELD_DIM 1
      #define FFH_T mytype_hashmap
      #define FFH_ENABLE_INT64
      #define FFH_EXISTING_KEY_IS_ERROR .true.
      #define FFH_INC_FILE "ffhash_inc.f90"
      #include "ffhash_typekey_inc.f90"

This fully hids the 'split-module' structure required for the use of `ffhash_inc.f90'.

> * I think it is good to have as few preprocessor statements as possible, 
which is why I did not include a `FFH_EXISTING_KEY_IS_ERROR`.

The reason is partially the same as above. 
In addition, it would be better if `existing_key_is_error = .true./.false.`  
wouldn't be a separate statement.
What about making `existing_key_is_error` an optional parameter 
in `store_key`?
Then we can immediately see how `store_key` is called.

> * The seed for a 64 bit hash function could also be 64 bit I guess.

In your `MurmurHash3_x64_128()` function, the seed is defined as `int32`, 
this is why I didn't change it.

Using a more effective hasher sounds interesting. 
PLease let me know what you find.

tobirolinew commented:
For compatibility reasons, I've removed `FFH_EXISTING_KEY_IS_ERROR` and switched back to `ffh_t` (the `unset` statement caused errors in a new procedure of my helper module, so I directly relplace `ffh_t` with a preprocessor variable). Thus, I think we only need `FFH_ENABLE_INT64` and `FFH_INT_KIND` as new variables for `int64` in `ffhash_inc.f90`.

jannisteunissen commented:
Thanks for your updates! I've had a look at a couple hash functions in the meantime. A general challenge is that there are no unsigned integer types (at least not with general compiler support) and that Fortran lacks some of the optimized integer arithmetic that you can do in C/C++. Here is a brief summary:

- xxHash: Difficult to port, pretty complex
- spookyhash: there is already a Fortran implementation (https://stdlib.fortran-lang.org/interface/spooky_hash.html) but that one is very slow. I think it is mostly due to the use of int8 but am not sure. (Other hash code from stdlib was also considerably slower than ffhash)
- cityhash: depends on byte swap methods, also a bit long
- mum hash: pretty short, I think I could port it, but good performance will depend on 128 bit integer support (AFAIK only in gfortran)
- rapidhash: compact, also has simplified versions. I ported the micro version (but still need to test). Fast with 128 bit integers, otherwise slower than my murmur3 implementation.

The good news is that the murmur3 Fortran code is not that slow, it achieves about 4 GB/s on my system (comparable to C version) and for small data it has a reasonable latency. So I think you can go ahead and create a pull request for the 64 bit hashing if you want.

tobirolinew commented:
Thanks for this thorough testing. I'll certainly create a new pull request, 
but first could you please help me address two problems 
(one is minor and the other is major -- both are related to custom types)?

(1) If I use this type as a key:

```
!***********************************************************************
      module mod_lpo
!***********************************************************************
      implicit none
      type :: lpo
         real*8 :: ener
         integer, allocatable :: coors(:)
         type(lpo), pointer :: parent
         class(*), pointer :: geo
      end type
      contains
      end
```

then I got a compiler error: 

```
The dummy argument of a derived type containing a pointer component 
has the INTENT(IN) attribute therefore it shall not appear as the expr of an 
assignment-stmt in a PURE procedure.   [KEY]
```

due to 

`FFH_KEY_ARG, intent(in)         :: key`

in `store_key` (this is caused by the pointer types present in it --- 
those are needed to generally treat a point in my project). 

If I replace the declaration of `key` is `store_key` to

`FFH_KEY_ARG, intent(inout)         :: key`

then the error disappears. Could we include this modification in `ffhash`?

---------------

(2) I'm trying to use your Murmur3 implementations for `FFH_KEY_FIELD` 
(which is `integer, allocatable :: coors(:)` in the custom type above) 
as a custom hash function implemented in the following way:

```
  pure function hash_function(key) result(hash)
    type(FFH_CUSTOM_TYPE), intent(in) :: key
    integer(int32), parameter :: seed = 0
    integer(int32) :: nbytes
    integer(INT_KIND) :: hash
#ifdef ENABLE_INT64
    integer(int32) :: hash128(4)
    integer(int64) :: h1, h2
#endif
    character(len=storage_size(0.125*key%FFH_KEY_FIELD)) :: buf
    buf = transfer(key%FFH_KEY_FIELD, buf)
#ifdef ENABLE_INT64
    call MurmurHash3_x64_128(buf, len(buf), seed, hash128)
    h1 = transfer(hash128(1:2), h1)
    h2 = transfer(hash128(3:4), h2)
    hash = ieor(h1, h2)
#else
    call MurmurHash3_x86_32(buf, len(buf), seed, hash)
#endif
  end function
```
  
However, this seems to be much slower (either with `int32` or `int64`) 
than just calling a primitive hasher for my special case, where the 
`FFH_KEY_FIELD` is a 1D integer array (again, `coors` above):

```
  pure integer(int64) function hash_function(key)
    implicit none
    type(FFH_CUSTOM_TYPE), intent(in) :: key
    integer :: i
    integer(int64) :: seed,c,elem
    c=int(z'9e3779b97f4a7c15',0_int64)
    seed=0_8
    do i=1,size(key%FFH_KEY_FIELD)
       elem=int(key%FFH_KEY_FIELD(i),0_int64)
       seed=ieor(seed,elem+c+ishft(seed,6)+ishft(seed,-2))
    end do
    hash_function=seed
  end function
```

I assume that this is because `integer`s can be hashed more efficiently 
than their `character`-type images.

I'm not sure how to treat this situation more generally, given the fact 
that Murmur3 is designed for strings.
Do you have an idea/suggestion for this?

jannisteunissen commented:
For (1), I would not change to `intent(inout)` but remove the `pure` attribute from the respective methods.

For (2): is the whole hash table much slower, or just the hash function itself when you benchmark it? I think hash functions like murmur3 (and also faster variants) will often be quite a bit slower on small keys than a simple function like you wrote. But for a large table that does not fit in the CPU cache, memory access will often be the bottleneck, and this might not matter so much.

Another thing: it would be nice to allow for an optional conversion function of the `key` type to the actual key used for hashing, which could e.g. extract fields from a user-defined type. Doing it generically is a bit hard, since the size of the key has to be known at compile time. But such custom hashing is actually already possible when using a custom hash function. So maybe we could simply say that for custom types with allocatables or pointers, a custom hash function should be used, and provide some templates?

tobirolinew commented:
Sorry for the delayed reply, and thanks for suggesting a test for the hash 
function’s execution time.
It turned out that the hashes were evaluated surprisingly fast, but only 
because the first entries of the integer vectors were hashed with

```
    character(len=storage_size(0.125*key%FFH_KEY_FIELD)) :: buf
    buf = transfer(key%FFH_KEY_FIELD, buf) 
```
which is, of course, incorrect.

To properly handle an array field, we need to account for the entire array size.
The difficulty is that this usually requires hard-coding the supported array ranks, 
unless we find a more elegant way of hashing arbitrary arrays (including those
specified as `character(:), allocatable`).

I’ve been experimenting with this, and it seems feasible to hash such arrays 
effectively using your `Murmur3` routines.
In my test cases, there’s no real difference in performance between the simple 
hasher and the Murmur3 implementation. 
(These tests involved ~900k integer vectors of length 12, with entries roughly 
between –10 and 10.)

I’m attaching my updated typekey helper, along with a test file 
([typekey_helper.zip](https://github.com/user-attachments/files/22020252/helper.zip)), 
where the key fields are `type%real` (0D), `type%character` (1D), and 
`type%integer` (2D). 
The hash values look sufficiently distinct, even in the 1D and 2D examples where 
certain coordinates have identical values.
In that code, the array dimensionality is allowed to be between 0 and 4.

I think it would be useful to provide **two dedicated helpers as complete Fortran 
modules**:

(1) **scalarkey_helper**: for intrinsic keys (`integer`, `real`, `logical`, and `character(:)`) 
(2) **typekey_helper**: for custom types, where a particular field of the key type 
is hashed.

In practice, the key field in (2) could take the form
```
FFH_KEY_FIELD = type%type2%type3%...%FIELD
```

where `FIELD` is either a scalar or array of `integer`, `real`, `logical`, or `character(:)` 
values.
This second case is the one I am trying to cover with my attached helper, since it’s 
the most relevant for my projects.

If you have suggestions on improving my implementation, I’d be glad to adjust it. 
A complete, self-contained module would also resemble C++ STL-style helpers, 
and may be easier to digest than the current split-module structure (which can be 
harder to follow for some users of this excellent `ffhash` library).

Finally, I’ll soon open a PR for the `int64` patch.
One additional point: could we make `existing_key_is_error` an optional argument 
in `store_key`? In a couple of my tests, I was confused when I forgot to set 
`ffht%existing_key_is_error = .true.`. 
And in fact, this flag is otherwise only used inside `store_key`.

jannisteunissen commented:
Let me reply in bullets:

- Good to see the performance of `Murmur3` is acceptable, that will save some effort.
- With a small performance penalty, it should also be possible to determine the key length at runtime. That would make hashing a lot more flexible.
- Regarding the two helpers: in principle, they have to do the same thing: take some data of known size (in whatever form, and perhaps flattened if necessary) and then pass it to a hash function. I haven't tried to implement this yet, but in principle this could be achieved by the introduction of an optional 'conversion' function that does this step.
- `existing_key_is_error` can indeed be an optional argument (for both `store_value` and `store_key`)

jannisteunissen commented:
Shall I give this idea about the conversion function a try? It is very similar to what you already included in your example, but just one step less (not calling the actual hash function).

I could imagine that in some applications, people want to combine multiple fields from a custom type. This will be a bit hard to do with preprocessor statements. But a very short conversion function to bytes could still do this generically.

jannisteunissen commented:
Here is an example: https://github.com/jannisteunissen/ffhash/tree/custom_convert_key

tobirolinew commented:
I really like your approach!
Having a `convert_key` function would indeed be very useful, 
since it would remove the need to handle the different cases 
through preprocessor variables. 
Writing such functions is not inconvenient at all—especially if 
the library provides a few typical examples. 
I’ll definitely try out this night how this would look in my project.

At first glance, it looks like the `#elif defined(FFH_KEY_IS_STRING)` 
branch can support free-length (`character(:), allocatable`) strings.
Is that correct?

Thanks again for your quick action and helpful thoughts!

tobirolinew commented:
I've tested this `convert_key` function and everything must be OK.
So I'm in favor of using this new custom procedure.

I think it would be useful to eliminate the `FFH_KEY_IS_STRING` and 
`FFH_VAL_IS_STRING` variables somehow.
Then it would be sufficient to set  only `FFH_KEY_TYPE` and  `FFH_VAL_TYPE`.

A way could be to define a default `convert_key` function as follows:

```
#ifndef FFH_CUSTOM_CONVERT_KEY
  pure function convert_key(key) result(buf)
    class(*), intent(in)      :: key
    character(:), allocatable :: buf
    select type(key)
    type is (character(*))
      buf = trim(key)
    class default
      buf = transfer(key, repeat(' ', storage_size(key)/8))
    end select
  end function convert_key
#endif
```

Then the default hasher would be only

```
  pure function hash_function(key) result(hash)
    FFH_KEY_ARG, intent(in) :: key
    integer                 :: hash
    integer, parameter      :: seed = 42
    call MurmurHash3_x86_32(convert_key(key), seed, hash)
  end function hash_function
```

It seems to work at first glance, but slighly (~2 times) slower 
than with static allocation for your benchmark example.
Using an allocatable buffer seems to be necessary, when the type 
is chosen via `select type`.

jannisteunissen commented:
Interesting idea to use select type, but the slowdown is a bit unfortunate.

I don't see a way to really eliminate the special flags for strings. (Un)fortunately strings are just quite "special" in Fortran, e.g. `"x " == "x"`. Without `FFH_VAL_IS_STRING` the code would emit warnings whenever the string passed as an argument is shorter than the maximum string length, e.g.:

```
#define FFH_VAL_TYPE character(len=15)
#include "ffhash_inc.f90"

...
25 |   call ha%ustore_value(1, "hello world")
      |                          1
Warning: Character length of actual argument shorter than of dummy argument ‘val’ (11/15) at (1)
```

An advantage of the current implementation is that you can also say that `"x" /= "x "` by not defining `FFH_KEY_IS_STRING` and using a custom compare function, that checks the string length. So I guess it will be difficult to support all of this with `select type`?

tobirolinew commented:
Yes, it’s a pity that `select type` tends to be slower, and you’re right that it’s not the most 
convenient solution.

You close your last comment with the following:

_An advantage of the current implementation is that you can also say that "x" /= "x " 
by not defining FFH_KEY_IS_STRING and using a custom compare function._

I don't think that it is a good idea not to define `FFH_KEY_IS_STRING` when `FFH_KEY` 
is a string.
However, if I'm correct, its value in the `#elif` block is not examined if 
`FFH_CUSTOM_CONVERT_KEY` is defined:

```
#if defined(FFH_CUSTOM_CONVERT_KEY)
    call MurmurHash3_x86_32(convert_key(key), seed, hash)
#elif defined(FFH_KEY_IS_STRING)
    call MurmurHash3_x86_32(trim(key), seed, hash)
#else
    integer, parameter      :: n_bytes = ceiling(storage_size(key)*0.125d0)
    character(len=n_bytes)  :: buf
    call MurmurHash3_x86_32(transfer(key, buf), seed, hash)
#endif
```

Thus, whenever `FFH_CUSTOM_CONVERT_KEY` is defined, that branch is always used, 
regardless of whether `FFH_KEY_IS_STRING` is also present. 
This is exactly the behavior most users would expect.

tobirolinew commented:
In addition to my previous ("shorter") comment, one last idea before I stop spamming you :D and 
move on with the `int64 hash`  PR:

How about defining `FFH_STRING_XXX_TYPE` instead of `FFH_XXX_IS_STRING` (`XXX = KEY or VAL`)? 
In that case,

```
#define FFH_KEY_TYPE character(len=20)
#define FFH_KEY_IS_STRING
#define FFH_VAL_TYPE character(len=15)
#define FFH_VAL_IS_STRING
```

could be simplified to 

```
#define FFH_STRING_KEY_TYPE character(len=20)
#define FFH_STRING_VAL_TYPE character(len=15)
```

on the user's side. 

Within `ffhash_inc.f90`, this would require initializing `FFH_KEY_TYPE` and `FFH_VAL_TYPE`, 
respectively, as

```
#ifdef FFH_STRING_KEY_TYPE
#  define FFH_KEY_TYPE FFH_STRING_KEY_TYPE
#endif
```

```
#ifdef FFH_STRING_VAL_TYPE
#  define FFH_VAL_TYPE FFH_STRING_VAL_TYPE
#endif
```

Then, we could use `FFH_STRING_KEY_TYPE` and `FFH_STRING_VAL_TYPE` in place of 
`FFH_KEY_IS_STRING` and `FFH_VAL_IS_STRING`, respectively. 
Of course, this wouldn’t affect the treatment of non-string keys/values.

Would this fit the editing style you prefer? 
If so, I can try to modify the code and the examples accordingly.

jannisteunissen commented:
I think that's an excellent idea! Shall I merge the `FFH_CUSTOM_CONVERT_KEY` part into master before you do your pull request, or is it okay if you include it in the pull request?

tobirolinew commented:
I'm glad you like this idea! I can start from your "custom_convert_key" branch and include:

(a) `FFH_STRING_KEY_TYPE` / `FFH_STRING_VAL_TYPE`,
(b) `existing_key_is_error` as an optional argument in `store_value` and `store_key`, and
(c) `FFH_ENABLE_INT64` for the `int64`-based hashing.

I'll also adapt your test codes to reflect point (a), and suggest an example case for 
`FFH_ENABLE_INT64`.

jannisteunissen commented:
That sounds perfect :)!
